networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/GossipMessageHandlerIntegrationTest.java

```diff
@@ -0,0 +1,98 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.p2p.jvmlibp2p;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.atLeastOnce;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.primitives.UnsignedLong;
+import java.util.Collection;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.network.p2p.jvmlibp2p.NetworkFactory;
+import tech.pegasys.artemis.networking.p2p.JvmLibP2PNetwork;
+import tech.pegasys.artemis.statetransition.BeaconChainUtil;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+import tech.pegasys.artemis.util.Waiter;
+
+public class GossipMessageHandlerIntegrationTest {
+
+  private final NetworkFactory networkFactory = new NetworkFactory();
+
+  @AfterEach
+  public void tearDown() {
+    networkFactory.stopAll();
+  }
+
+  @Test
+  public void shouldGossipBlocksAcrossToIndirectlyConnectedPeers() throws Exception {
+    // Setup network 1
+    final EventBus eventBus1 = new EventBus();
+    final ChainStorageClient storageClient1 = new ChainStorageClient(eventBus1);
+    final JvmLibP2PNetwork network1 = networkFactory.startNetwork(eventBus1, storageClient1);
+    final BeaconChainUtil chainUtil = BeaconChainUtil.create(12, storageClient1);
+
+    // Setup network 2
+    final EventBus eventBus2 = spy(new EventBus());
+    final ChainStorageClient storageClient2 = new ChainStorageClient(eventBus2);
+    final JvmLibP2PNetwork network2 = networkFactory.startNetwork(eventBus2, storageClient2);
+    chainUtil.initializeStorage(storageClient2);
+
+    // Setup network 3
+    final EventBus eventBus3 = spy(new EventBus());
+    final ChainStorageClient storageClient3 = new ChainStorageClient(eventBus3);
+    final JvmLibP2PNetwork network3 = networkFactory.startNetwork(eventBus3, storageClient3);
+    chainUtil.initializeStorage(storageClient3);
+
+    // Connect networks 1 -> 2 -> 3
+    network1.connect(network2.getPeerAddress());
+    network2.connect(network3.getPeerAddress());
+    // Wait for connections to get set up
+    Waiter.waitFor(
+        () -> {
+          assertThat(network1.getPeerManager().getAvailablePeerCount()).isEqualTo(1);
+          assertThat(network2.getPeerManager().getAvailablePeerCount()).isEqualTo(2);
+          assertThat(network3.getPeerManager().getAvailablePeerCount()).isEqualTo(1);
+        });
+    // TODO: debug this - we shouldn't have to wait here
+    Thread.sleep(2000);
```

Comment:
I hit this as well - it looks like libp2p considers the peer connected and we can exchange STATUS messages before it actually completes subscribing to the gossip channels.  I'd love to have a way to know when that subscription is completed.  @shahankhatch from your libp2p work, do you know of anyway to know when a peer's channels are all fully established?